api/server/spanconverter.go

```diff
@@ -0,0 +1,129 @@
+package server
+
+import (
+	"context"
+	"strings"
+	"sync"
+	"time"
+	"unicode"
+
+	"go.opencensus.io/stats"
+	view "go.opencensus.io/stats/view"
+	"go.opencensus.io/trace"
+)
+
+// SpanConverter registers as a opencensus Trace Exporter,
+// but it converts all the Spans in Views and registers them as such
+// A View exporter will then export them as normal.
+type SpanConverter struct {
+	opts     Options
+	measures map[string]*stats.Float64Measure
+	viewsMu  sync.Mutex
+	e        view.Exporter
+}
+
+// Options contains options for configuring the exporter.
+type Options struct {
+	Namespace string
+}
+
+func NewSpanConverter(o Options) (*SpanConverter, error) {
+	c := &SpanConverter{
+		opts:     o,
+		measures: make(map[string]*stats.Float64Measure),
+	}
+	return c, nil
+}
+
+var maxViews = 100
+
+// Spans are rejected if there are already maxViews (100) or they are
+// prefixed with '/', gin as been observed creating Span id specific
+// named Spans.
+func (c *SpanConverter) rejectSpan(sd *trace.SpanData) bool {
+	return len(c.measures) > maxViews || urlName(sd)
+}
+
+// ExportSpan creates a Measure and View once per Span.Name, registering
+// the View with the opencensus register. The length of time reported
+// by the span is then recorded using the measure.
+func (c *SpanConverter) ExportSpan(sd *trace.SpanData) {
+	if c.rejectSpan(sd) {
+		return
+	}
+	m := c.getMeasure(sd)
+
+	spanTimeNanos := sd.EndTime.Sub(sd.StartTime)
+	spanTimeMillis := float64(int64(spanTimeNanos / time.Millisecond))
+
+	stats.Record(context.Background(), m.M(spanTimeMillis))
```

Comment:
we should rip some of the attributes off the span data as well and put them onto the context. this may be slightly painful, but tags are likely important so that we have context for eg function_id, app_id... these seem to map to the fields `SpanData.Attributes` and `tag.Map`, however it's not immediately clear from digging around the docs.

Reply:
Attributes, MessageEvents and Annotations are all empty for our spans.